libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -449,44 +666,116 @@ BeaconSetupService::State BeaconSetupService::OnWaitForReconstructionShares()
         beacon_->manager.VerifyReconstructionShare(from, elem);
       }
     }
+
     if (!beacon_->manager.RunReconstruction())
     {
       FETCH_LOG_WARN(LOGGING_NAME, "Node: ", beacon_->manager.cabinet_index(),
-                     " DKG failed due to reconstruction failure");
+                     " DKG failed due to reconstruction failure. Resetting.");
+      SetTimeToProceed(State::RESET);
+      return State::RESET;
     }
     else
     {
       beacon_->manager.ComputePublicKeys();
     }
-    return State::BEACON_READY;
+
+    SetTimeToProceed(State::DRY_RUN_SIGNING);
+    return State::DRY_RUN_SIGNING;
   }
+
   state_machine_->Delay(std::chrono::milliseconds(10));
   return State::WAIT_FOR_RECONSTRUCTION_SHARES;
 }
 
+/**
+ * Attempt to sign the seed to determine enough peers have generated the
+ * group public signature
+ */
+BeaconSetupService::State BeaconSetupService::OnDryRun()
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  // Only on first entry to this function
+  if (state_machine_->previous_state() != State::DRY_RUN_SIGNING)
+  {
+    beacon_->manager.SetMessage("test message");
+    beacon_->member_share = beacon_->manager.Sign();
+
+    // insert ourselves - others will insert here also via gossip
+    dry_run_shares_[identity_.identifier()] = beacon_->member_share;
+
+    DryRunInfo to_send{beacon_->manager.group_public_key(), beacon_->member_share};
+
+    // Gossip this to everyone
+    {
+      fetch::serializers::SizeCounter counter;
+      counter << to_send;
+
+      fetch::serializers::MsgPackSerializer serializer;
+      serializer.Reserve(counter.size());
+      serializer << to_send;
+      endpoint_.Broadcast(SERVICE_DKG, CHANNEL_SIGN_DRY_RUN, serializer.data());
+    }
+  }
+
+  if (timer_to_proceed_.HasExpired())
+  {
+    bool found_key = false;
+
+    for (auto const &key_and_count : dry_run_public_keys_)
+    {
+      if (key_and_count.second >= beacon_->manager.polynomial_degree())
+      {
+        found_key = true;
+      }
+    }
+
+    if (!found_key)
+    {
```

Comment:
I think the check here should be that 1) Is consensus reach on any key and then 2) whether this key is the group public key the node has computed. If 1) fails then everyone should reset the DKG and if 2) fails then the node has just failed this particular DKG instance (but shouldn't reset since others have suceeded). In either case of failure, the node doesn't need to bother with going on to computing the group signature

Reply:
Ok, I have changed it as you suggest, to this:

```
  if (timer_to_proceed_.HasExpired())
  {
    bool found_key     = false;
    bool found_our_key = false;

    for (auto const &key_and_count : dry_run_public_keys_)
    {
      if (key_and_count.second >= beacon_->manager.polynomial_degree())
      {
        found_key     = true;
        found_our_key = beacon_->manager.group_public_key() == key_and_count.first;
      }
    }

    if (!found_key)
    {
      FETCH_LOG_WARN(LOGGING_NAME, "Failed to reach consensus on group public key!");
      SetTimeToProceed(State::RESET);
      return State::RESET;
    }

    if (!found_our_key)
    {
      FETCH_LOG_WARN(LOGGING_NAME, "Other nodes didn't agree with our computed group public key!");
      SetTimeToProceed(State::RESET);
      return State::RESET;
    }

    for (auto const &share : dry_run_shares_)
    {
      beacon_->manager.AddSignaturePart(share.second.identity, share.second.signature);
    }

    bool const could_sign = beacon_->manager.can_verify() && beacon_->manager.Verify();

    if (could_sign)
    {
      SetTimeToProceed(State::BEACON_READY);
      return State::BEACON_READY;
    }
    else
    {
      FETCH_LOG_WARN(LOGGING_NAME, "Failed to complete dry run for signature signing!");
      SetTimeToProceed(State::RESET);
      return State::RESET;
    }
  }
```

Note that it is fine for failed nodes to reset since once they see a valid block being generated from the other members of the cabinet they will call abort on the DKG setup.